src/nupic/encoders/ScalarEncoder.hpp

```diff
@@ -21,167 +23,119 @@
   };
 
-/** Encodes a floating point number as a contiguous block of 1s.
- *
- * @b Description
- * A ScalarEncoder encodes a numeric (floating point) value into an array
- * of bits. The output is 0's except for a contiguous block of 1's. The
- * location of this contiguous block varies continuously with the input value.
- *
- * Conceptually, the set of possible outputs is a set of "buckets". If there
- * are m buckets, the ScalarEncoder distributes m points along the domain
- * [minValue, maxValue], including the endpoints. To figure out the bucket
- * index of an input, it rounds the input to the nearest of these points.
- *
- * This approach is different from the PeriodicScalarEncoder because two
- * buckets, the first and last, are half as wide as the rest, since fewer
- * numbers in the input domain will round to these endpoints. This behavior
- * makes sense because, for example, with the input space [1, 10] and 10
- * buckets, 1.49 is in the first bucket and 1.51 is in the second.
- */
-class ScalarEncoder : public ScalarEncoderBase {
-public:
   /**
-   * Constructs a ScalarEncoder
-   *
-   * @param w The number of bits that are set to encode a single value -- the
-   *   "width" of the output signal
-   * @param minValue The minimum value of the input signal, inclusive.
-   * @param maxValue The maximum value of the input signal, inclusive.
-   * @param clipInput Whether to allow input values outside the [minValue,
-   * maxValue] range.  If true, the input will be clipped to minValue or
-   * maxValue.
+   * Encodes a real number as a contiguous block of 1s.
    *
-   * There are three mutually exclusive parameters that determine the overall
-   * size of of the output. Only one of these should be nonzero:
-   *
-   * @param n The number of bits in the output. Must be greater than or equal to
-   * w.
-   * @param radius Two inputs separated by more than the radius have
-   *   non-overlapping representations. Two inputs separated by less than the
-   *   radius will in general overlap in at least some of their bits. You can
-   *   think of this as the radius of the input.
-   * @param resolution Two inputs separated by greater than, or equal to the
-   *   resolution are guaranteed to have different representations.
+   * Description:
+   * The ScalarEncoder encodes a numeric (floating point) value into an array
+   * of bits. The output is 0's except for a contiguous block of 1's. The
+   * location of this contiguous block varies continuously with the input value.
    */
-  ScalarEncoder(int w, double minValue, double maxValue, int n, double radius,
-                double resolution, bool clipInput);
+  class ScalarEncoder : public BaseEncoder<double>
+  {
+  public:
+    ScalarEncoder() {};
+    ScalarEncoder( ScalarEncoderParameters &parameters );
 
-  int encodeIntoArray(Real input, UInt output[]) override;
+#ifdef __clang__
+  #pragma clang diagnostic push
+  #pragma clang diagnostic ignored "-Woverloaded-virtual"
```

Comment:
I'm not a fan of these, but ok if there's no better way

Reply:
That was my reaction too.  The problem is that we use `-Wall -Wextra -Werror` and so overzealous compilers try to stop us from making any mistakes, like using C++  :)

Reply:
Have you read https://stackoverflow.com/questions/18515183/c-overloaded-virtual-function-warning-by-clang ? 
Esp solutions 1, 3?

Reply:
Fixed.  I made the method `BaseEncoder::initialize` not virtual, since it needs to be directly called and should not be overridden.

Reply:
thanks! i'm happy with the better approach